### DISCUSSION TITLE
Does Django-Ratelimit clear its own keys?

### DISCUSSION THREAD

**Discussion by necrolingus** [jsocol/django-ratelimit#235] [Q&A] [answered]
Good day

I Googled for this but could not seem to find any answer. 
Let's say for example that I use file based caching for django-ratelimit, I found that manage.py clearsessions doesn't clear up any of the files django-ratelimit created, even hours after they have been created (rate limit is set to e.g. 2/m or 1/h). 
Will django-ratelimit just continue to add files with no real way to clean them up?

And then the same question for e.g. memcached/Redis. 
Will django-ratelimit just continue to add keys? If so, I guess it is up to the way memcached/Redis is set up to delete the keys, e.g. using LRU because it is not possible to set a TTL for keys created by django-ratelimit?

I forgot to add I am using django-ratelimit 3.0.1, django 2.2.19 running on Windows Server 2012 R2.

Thank you

**Comment by jsocol** [comment, marked as answer]
Checking the code, `django-ratelimit` does [set expiration times](https://github.com/jsocol/django-ratelimit/blob/main/django_ratelimit/core.py#L224) on the keys it creates so that they should last slightly longer than the required window.

The filesystem cache doesn't implement expiration or LRU, but it does obey the [`CULL_FREQUENCY` and `MAX_ENTRIES` options](https://docs.djangoproject.com/en/3.1/topics/cache/#cache-arguments). But since eviction is random, rather than LRU, that may not be the best choice.

Technically `django-ratelimit` should work with any Django cache backend that implements atomic increment—memcached, Redis, and even the `LocMem` backends do, but the filesystem backend does not. Ratelimit is designed assuming situations where you're likely running an application on multiple hosts and will need a shared caching setup.

**Reply by necrolingus** [reply]
Thanks for the detailed reply jsocol, you answered my question 100% and gave me clarity.
